File: microservice/microservice-app/src/main/java/com/greg/microservice/web/TextDoublerController.java
package com.greg.microservice.web;

import com.greg.microsservice.shared.client.TextDoublerClient;
import com.greg.microsservice.shared.model.TextDoubleRequest;
import com.greg.microsservice.shared.model.TextDoubleResult;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;

@RestController
@RequestMapping("/home")
@Slf4j
public class TextDoublerController implements TextDoublerClient {
    @PostMapping("text-double")
    public TextDoubleResult hello(@Valid @RequestBody TextDoubleRequest textDoubleRequest) {
        log.info("Message recieved "+ textDoubleRequest.text());
        return new TextDoubleResult(textDoubleRequest.text() + textDoubleRequest.text());
    }

    @GetMapping("test-error")
    public void testError() {
        var customer = new Customer("Bob",new Address(new Country(null)));
        log.info(customer.address().country().timeZone().zone());
    }

    @GetMapping("delay")
    public void delay(@RequestParam("milliseconds") int milliseconds)  {
        try {
            Thread.sleep(milliseconds);
        } catch (InterruptedException e) {
            throw  new RuntimeException(e);
        }
    }

    record Customer(String  name, Address address){

    }

    record Address(Country country){

    }

    record Country(TimeZone timeZone){

    }

    record TimeZone(String zone){

    }
}

File: microservice/microservice-app-kotlin/src/main/kotlin/com/greg/microserviceappkotlin/web/TextDoublerController.kt
package com.greg.microserviceappkotlin.web

import com.greg.microsservice.shared.client.TextDoublerClient
import com.greg.microsservice.shared.model.TextDoubleRequest
import com.greg.microsservice.shared.model.TextDoubleResult
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.web.bind.annotation.*
import java.lang.InterruptedException
import java.lang.RuntimeException
import javax.validation.Valid

@RestController
@RequestMapping("/home")
class TextDoublerController : TextDoublerClient {
    private val log: Logger = LoggerFactory.getLogger(this::class.java)
    @PostMapping("text-double")
    override fun hello(@RequestBody textDoubleRequest: @Valid TextDoubleRequest): TextDoubleResult {
        log.info("Message recieved " + textDoubleRequest.text())
        return TextDoubleResult(textDoubleRequest.text() + textDoubleRequest.text())
    }

    @GetMapping("test-error")
    override fun testError() {
        val num = 10/0;
        return;
    }

    @GetMapping("delay")
    override fun delay(@RequestParam("milliseconds") milliseconds: Int) {
        try {
            Thread.sleep(milliseconds.toLong())
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }
}